File: 76-minimum-window-substring/76-minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        m, n = len(s), len(t)
        if m < n:
            return ""
        left, right = 0, 0
        start, l = 0, float('inf')
        needs = collections.Counter(t)
        window = collections.defaultdict(int)
        valid = 0
        while right < m:
            c = s[right]
            right += 1
            if c in needs:
                window[c] += 1
                if needs[c] == window[c]:
                    valid += 1
                
            while valid == len(needs):
                if right - left < l:
                    start = left
                    l = right - left
                d = s[left]
                left += 1
                if d in needs:
                    if window[d] == needs[d]:
                        valid -= 1
                    window[d] -= 1
        return "" if l > m else s[start:start + l]

File: 76-minimum-window-substring/76-minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {
        int m = s.length(), n = t.length();
        if (m < n) {
            return "";
        }
        int left = 0, right = 0;
        int start = 0, l = Integer.MAX_VALUE;
        HashMap<Character, Integer> needs = new HashMap<Character, Integer>();
        for (int i = 0; i < n; i++) {
            char c = t.charAt(i);
            int count = needs.getOrDefault(c, 0);
            needs.put(c, count + 1);
        }
        HashMap<Character, Integer> window = new HashMap<Character, Integer>();
        int valid = 0;
        while (right < m) {
            char c = s.charAt(right);
            right++;
            if (needs.containsKey(c)) {
                int count = window.getOrDefault(c, 0);
                window.put(c, count + 1);
                if (window.get(c).intValue() == needs.get(c).intValue()) {
                    valid++;
                }
            }
            //System.out.printf("%s, %s\n", start, l);
            while (valid == needs.size()) {
                if (right - left < l) {
                    start = left;
                    l = right - left;
                }
                char d = s.charAt(left);
                left++;
                if (needs.containsKey(d)) {
                    if (needs.get(d).intValue() == window.get(d).intValue()) {
                        valid -= 1;
                    }
                    window.put(d, window.get(d) - 1);
                }
            }
        }
        return l == Integer.MAX_VALUE ? "" : s.substring(start, start + l);
    }
}